src/AutoRest.CSharp/Mgmt/AutoRest/PostProcess/Internalizer.cs

```diff
@@ -139,22 +139,18 @@ private static bool HasDiscriminator(BaseTypeDeclarationSyntax node, [MaybeNullW
             // only class models will have discriminators
             if (node is ClassDeclarationSyntax classDeclaration)
             {
-                var methodName = $"Deserialize{classDeclaration.Identifier.Text}";
-                var deserializeMethod = classDeclaration.Members.OfType<MethodDeclarationSyntax>().FirstOrDefault(method => method.Identifier.Text.Equals(methodName));
-                if (deserializeMethod != null && deserializeMethod.Body != null)
+                if (classDeclaration.HasLeadingTrivia)
                 {
-                    var switchStatementNode = deserializeMethod.Body.DescendantNodes().OfType<SwitchStatementSyntax>();
-                    foreach (var switchNode in switchStatementNode)
+                    var syntaxTriviaList = classDeclaration.GetLeadingTrivia();
+                    var filteredTriviaList = syntaxTriviaList.Where(syntaxTrivia => BuilderHelpers.DiscriminatorDescFixedPart.All(syntaxTrivia.ToFullString().Contains));
+                    if (filteredTriviaList.Count() == 1)
                     {
-                        if (switchNode.Expression.ToFullString().Contains("discriminator.GetString"))
-                        {
-                            // get the identifiers in the switch statements, some of them are class names of the derived classes
-                            // but since we do not have the full semantic model here, we cannot directly get which of them is
-                            // therefore here we just return all of them, and see if the derived type name shows in this list
-                            var identifierNodes = switchNode.Sections.SelectMany(section => section.DescendantNodes().OfType<IdentifierNameSyntax>());
-                            identifiers = identifierNodes.Select(identifier => identifier.Identifier.ToFullString()).ToImmutableHashSet();
-                            return true;
-                        }
+                        var descendantNodes = filteredTriviaList.First().GetStructure()?.DescendantNodes().ToList();
+                        var filteredDescendantNodes = descendantNodes.Where((val, index) =>
+                            index >= descendantNodes?.FindLastIndex(node => node.ToFullString().Contains(BuilderHelpers.DiscriminatorDescFixedPart.Last())));
```

Comment:
This nested linq is really hard to understand, can we simplify this by adding some method for this?

Reply:
Good point, I added a separate method to replace this linq. And in the method also some comments are added.